components/notebook-controller/pkg/controller/notebook/notebook_controller.go

```diff
@@ -123,18 +122,81 @@ func (r *ReconcileNotebook) Reconcile(request reconcile.Request) (reconcile.Resu
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
-	if err = r.ReconcileStatefulSet(instance); err != nil {
+
+	// Reconcile StatefulSet
+	ss := generateStatefulSet(instance)
+	if err := controllerutil.SetControllerReference(instance, ss, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
-	if err = r.ReconcileService(instance); err != nil {
+	// Check if the StatefulSet already exists
+	foundStateful := &appsv1.StatefulSet{}
+	justCreated := false
+	err = r.Get(context.TODO(), types.NamespacedName{Name: ss.Name, Namespace: ss.Namespace}, foundStateful)
+	if err != nil && errors.IsNotFound(err) {
+		log.Info("Creating StatefulSet", "namespace", ss.Namespace, "name", ss.Name)
+		err = r.Create(context.TODO(), ss)
+		justCreated = true
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if err != nil {
 		return reconcile.Result{}, err
 	}
+	// Update the foundStateful object and write the result back if there are any changes
+	if !justCreated && util.CopyStatefulSetFields(ss, foundStateful) {
+		log.Info("Updating StatefulSet", "namespace", ss.Namespace, "name", ss.Name)
+		err = r.Update(context.TODO(), foundStateful)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Reconcile service
+	service := generateService(instance)
+	if err := controllerutil.SetControllerReference(instance, service, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+	// Check if the Service already exists
+	foundService := &corev1.Service{}
+	justCreated = false
+	err = r.Get(context.TODO(), types.NamespacedName{Name: service.Name, Namespace: service.Namespace}, foundService)
+	if err != nil && errors.IsNotFound(err) {
+		log.Info("Creating Service", "namespace", service.Namespace, "name", service.Name)
+		err = r.Create(context.TODO(), service)
+		justCreated = true
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if err != nil {
+		return reconcile.Result{}, err
+	}
+	// Update the foundService object and write the result back if there are any changes
+	if !justCreated && util.CopyServiceFields(service, foundService) {
+		log.Info("Updating Service\n", "namespace", service.Namespace, "name", service.Name)
+		err = r.Update(context.TODO(), foundService)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Update the status if previous condition is not "Ready"
+	oldConditions := instance.Status.Conditions
+	if len(oldConditions) == 0 || oldConditions[0].Type != "Ready" {
```

Comment:
Is the last condition at index 0 or -1?

Reply:
I am doing prepending, (line 188), so should be 0.
I think prepending is convention, e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L136